TestResultSummaryService/routes/postIssueFeedback.js

```diff
@@ -0,0 +1,159 @@
+            .toArray();
+        return {
+            ...testResultData,
+            testOutput: result[0].output,
+        };
+    } catch (error) {
+        throw error;
+    }
+}
+
+async function writeTestOutputToFile(
+    testName,
+    buildName,
+    buildUrl,
+    buildNum,
+    testOutput
+) {
+    try {
+        const domain = new URL(buildUrl).hostname;
+        const fileName = `${testName}_${buildName}_${buildNum}_${domain}.txt`;
+        const outputPath = `${__dirname}/../../MachineLearningPrototype/data/JenkinsDataWithFeedback/${fileName}`;
+
+        const fileExists = await fs.stat(outputPath).then(
+            () => true,
+            () => false
+        );
+
+        if (!fileExists) {
+            await fs.writeFile(outputPath, testOutput);
+        }
+
+        return fileName;
+    } catch (error) {
+        throw error;
+    }
+}
+
+async function getRecordIfExists(db, repoName, issueNumber, testId) {
+    const records = await db
+        .getData({ repoName, issueNumber, testId })
+        .toArray();
+
+    if (records.length == 0) {
+        return null;
+    }
+
+    return records[0];
+}
+
+async function insertNewFeedback(db, requestBody) {
+    const {
+        repoName,
+        issueName,
+        issueCreator,
+        testId,
+        testName,
+        buildName,
+        issueNumber,
+        accuracy,
+    } = requestBody;
+
+    const { testOutput, buildUrl, buildNum } = await getTestOutput(testId);
+    const outputFileName = await writeTestOutputToFile(
+        testName,
+        buildName,
+        buildUrl,
+        buildNum,
+        testOutput
+    );
+
+    const positiveCounter = Number(accuracy === 1);
+    const negativeCounter = Number(accuracy === 0);
+
+    await db.populateDB({
+        repoName,
+        issueName,
+        issueCreator,
+        issueNumber,
+        testName,
+        testId,
+        testOutput: outputFileName,
+        positiveCounter,
+        negativeCounter,
+    });
+}
+
+async function updateFeedback(db, record, accuracy) {
+    const options = { upsert: false };
+    const criteria = { _id: new ObjectID(record._id) };
+
+    let { positiveCounter, negativeCounter } = record;
+
+    if (accuracy === 0) {
+        negativeCounter += 1;
+    }
+
+    if (accuracy === 1) {
+        positiveCounter += 1;
+    }
+
+    await db.update(
+        criteria,
+        { $set: { positiveCounter, negativeCounter } },
+        options
+    );
+}
+
+module.exports = async (req, res) => {
+    try {
+        const db = new FeedbackDB();
+        const { repoName, testId, issueNumber, accuracy } = req.body;
+
+        const record = await getRecordIfExists(
+            db,
+            repoName,
+            issueNumber,
+            testId
+        );
+
+        if (!record) {
+            await insertNewFeedback(db, req.body);
+        } else {
+            await updateFeedback(db, record, accuracy);
+        }
```

Comment:
I think we should use `upsert`

Reply:
Hi @llxia, if we want to upsert we may have to query for the testOutput regardless of whether it's an insert or update operation. Branching them based on the record's existence may save us one separate database query. Although I have already implemented upsert on my local branch I thought I should ask what you think about this. Thanks.

Reply:
In upsert, we should be able to know to insert or update based on the return value.
testOutput is needed if the file does not exist on disk. Having the record in db does not mean that the testOutput is stored on disk.

Reply:
Mmm, I think I am missing something here. Are you talking to replace updateFeedback() with an upsert operation? What I inferred last time is to remove the check for the record's existence [getRecordIfExists()] and directly upsert the input rather than using two methods (insertNewFeedback, updateFeedback).